28 infra integrate markdown link checkin

sd-ditoy opened the pull request:
### Related links

https://github.com/casper-network/docs/issues/28 INFRA: integrate markdown link checking module

### Changes

- created standalone script that retrieves all external URLs from all available markdown files
- uses puppeteer to check for url validity
- results will be output in the action log and if there are invalid urls found, it will fail the build process
- results includes FAIL state, for broken links. WARN states for links that may break in the future and INFO states, which as the name implies just informs you of possible future issues. Only FAILED checks will fail the build.

### Notes

As noted, the idea of this check is to fail the build process. as it stands right now there are already invalid
links in the documentation markdown files. So be aware that when this would be merged to main, any
further builds will brick until the links are corrected.

Since @ipopescu will be on holidays, i was told to add @Buddhika-Pathirathna , @GloreenS or @ACStoneCL 
as reviewers 👋

ACStoneCL commented:
Most of the failed links appear to be in the /i18n/ directory, which I believe is the basis for our other language documentation - and not currently up to date. I could be wrong on that.

I checked the other two failed links, and I can rectify the issue in the Developers FAQ - but the [URL from the ERC20 Tutorial](https://github.com/casper-ecosystem/erc20/blob/master/example/erc20-tests/src/test_fixture.rs) appears valid, yet this is marking it as a 404. 

For the WARN issues, it might be worth opening a new docs card to go through and correct/amend the URLs as necessary.

sd-ditoy commented:
> Most of the failed links appear to be in the /i18n/ directory, which I believe is the basis for our other language documentation - and not currently up to date. I could be wrong on that.
> 
> I checked the other two failed links, and I can rectify the issue in the Developers FAQ - but the [URL from the ERC20 Tutorial](https://github.com/casper-ecosystem/erc20/blob/master/example/erc20-tests/src/test_fixture.rs) appears valid, yet this is marking it as a 404.
> 
> For the WARN issues, it might be worth opening a new docs card to go through and correct/amend the URLs as necessary.

I could add folder exclusion if that is wished for. Are there any other folders you think we should exclude from that check?

Regarding the ERC20 Tutorial Link - that link is indeed incorrect. In the documentation there is an addtional ">" character in the URL, which can cause issues. The link appears twice in the same file and one of them contains that character by mistake it seems.

ACStoneCL commented:
I see what you mean about the extra ">". I went through and rectified all of the link issues in our main /docs/ directory. [Here's the pull request](https://github.com/casper-network/docs/pull/224).

I think we should exclude the /i18n/ folder for the time being, but I also think it would be best to have @Buddhika-Pathirathna or @GloreenS weigh in before making any final decisions on it.

sd-ditoy commented:
Ok. I will update the code to add a way to exclude folders. From what i see currently, i will exclude `./source/i18n` and `.source/blog`.

So let's wait with any approval until i've pushed those changes. Definitely wouldn't mind others weighing in, in the meantime. 👍

sd-ditoy commented:
I made some small improvements and think this is now ready to be fully reviewed:

- added folder exclusion as we discussed
- renamed the yarn action to align/group with the other checker that has also been worked on: https://github.com/casper-network/docs/tree/125-clean-up-unused-files
- i increased the url request interval since github has rate limiting, which we've been hitting 😄

sd-ditoy commented:
I've resolved the merge conflict that was open with `package.json` and all neccessary checks seem to pass now. I think we are good to go, unless there are any other open concerns?

ipopescu commented:
Thanks for working on this, @sd-ditoy and reviewers! Can we merge this?
